docs/RFCS/20230731_region_liveness.md

```diff
@@ -0,0 +1,363 @@
+`system.sqlliveness` sessions will have expired by the time the region is
+considered unavailable.
+
+When a server wishes to create or extend a `system.sqlliveness` session, it
+must consult the `system.region_liveness` table. If `unavailable_at` IS NOT
+NULL, then the server must pick a `sqlliveness` expiration timestamp that comes
+before the `unavailable_at` timestamp.
+
+If `unvailable_at` is not null, but set to a timestamp after the transaction
+timestamp, a SQL server can attempt to set it to null in a transaction
+modifying only the region_liveness table. If `unavailable_at` is before the
+transaction timestamp, then the server must delete all rows in the
+`system.sqlliveness`, `system.sql_instances`, and `system.lease` tables in the
+transaction that sets `unavailable_at` to `NULL`. The deletion is necessary
+because servers in other regions may have acted as if the rows in the region
+did not exist.
+
+# Details
+
+## Region Liveness
+
+### How is the `unavailable_at` timestamp picked?
+
+When writing to the `unavailable_at` timestamp, an expiration time must be
+picked that is far enough in the future that all `system.sqlliveness` will have
+expired by the time the region is treated as unavailable. This decision must be
+made without reading from the `system.sql_instances` table, because if the
+region is unavailable, then reading the unavailable region's rows in the
+`system.sql_instances` table will timeout. 
+
+The simplest option is to pick an `unavailable_at` timestamp that is the
+transaction timestamp +
+[server.sqlliveness.ttl](https://github.com/cockroachdb/cockroach/blob/37e61e9e245ac623b461e7917761c27d0223f1bb/pkg/sql/sqlliveness/slinstance/slinstance.go#L42).
+It’s not necessary to account for clock skew when picking `unavailable_at`
+because the logic is based on transaction timestamp and the SQL servers writing
+to the `system.sqlliveness` table are reading from the `system.region_liveness`
+table in a transaction. 
+
+There is one danger in picking the `unavailable_at` timestamp based on the
+`system.sqlliveness.ttl`, which is the `system.sqlliveness.ttl`  is a setting
+that users may change.
+
+Consider:
+T0: a session is created in region-a that expires at T0+30=T30
+T1: the session expiration setting is changed to +5 seconds.
+T2: region-a is marked as expired at T0+5=T5. 
+
+In this example, the region-a will be treated as expired at T5, while there is
+still an active session in the region, which violates the
+`system.region_liveness` table’s invariants. 
+
+Here are a few ways the risk of changing the `system.sqlliveness.ttl` setting
+could be mitigated.
+* Option 1: Look at all recent values of the setting and pick the max value when
+  fencing a region. This is the most work, but the infrastructure for reading
+  multiple versions of a setting would also be useful for improving the behavior
+  of cluster setting version gates.
+* Option 2: Pick something like 2x the setting value as the region expiration
```

Comment:
I think this one won't work unless we make sure it propogates somehow?

Reply:
Couldn't a user simply double the value twice to re-create the problem or would this be a constraint enforced at commit time? Even if it would be the latter, I don't think I understand how this is an effective mitigation?

Reply:
I added a caveat to this bullet point. I view the validation approaches as a bit of hack, but probably good enough in practice. After all it requires multiple setting changes to make unsafe changes and the unsafety only crops up if a region is marked unavailable around the time of the setting change.